#164139767 User Registration with JWT

LordUche opened the pull request:
#### What does this PR do?
Sends a JWT token on successful user registration
#### Description of Task to be completed?
- write unit test
- create route
- validate inputs with `joi`
- create controller
#### How should this be manually tested?
- ensure that the build is passing on travis 
- test the `POST /api/users` endpoint with Postman at the [review app](https://vale-ah-backend-staging-pr-12.herokuapp.com/)
#### Any background context you want to provide?
Users once registered via the registration endpoint, should receive a JWT to be required on all subsequent calls to all other endpoints that require authentication
#### What are the relevant pivotal tracker stories?
[#164139767](https://www.pivotaltracker.com/story/show/164139767)
#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

LordUche committed:
feat(registration): user registration

- write unit test
- create route
- validate inputs with joi
- create controller

[Delivers #164139767]

blackdevelopa reviewed (commented):
This is impressive. You took so much time though.

devskope commented on server/routes/index.js:
```diff
@@ -1,22 +1,13 @@
-});
+const router = (app) => {
+  app.use('/api-docs', serve, setup(apiSpec));
+  app.use('/api/users', usersRoutes);
```
`usersRoutes` should not be pluralized. 

A single router should suffice for routing all user related endpoints to appropriate controllers.

devskope commented on server/controllers/users-controller.js:
```diff
@@ -0,0 +1,40 @@
+ *
+ * @class UsersController
+ */
+class UsersController {
```
`UsersController` should not be pluralized.
The handler methods contained within should be linked to all user related routes

andela-pessien replied:
Pluralization (or lack of) is pretty much a matter of preference here.

devskope commented on server/controllers/users-controller.js:
```diff
@@ -0,0 +1,40 @@
+     * @returns {undefined}
+     */
+  static register(req, res) {
+    const { user } = req.body;
```
The fields of interest should ideally be directly on the `request.body` object.

andela-pessien replied:
Agreed

LordUche replied:
From the `README`, file the field of interest is `user` 👇 
```
{
  "user":{
    "username": "Jacob",
    "email": "jake@jake.jake",
    "password": "jakejake"
  }
}
```
Am I permitted to change the API spec and hence the README?

andela-pessien commented on package.json:
```diff
@@ -19,20 +18,24 @@
   "license": "MIT",
   "dependencies": {
     "@babel/polyfill": "^7.2.5",
+    "bcrypt": "^3.0.4",
```
Ideally you should pin down dependency versions much tighter than this.

andela-pessien commented on server/middlewares/users-middleware.js:
```diff
@@ -0,0 +1,25 @@
+      * @memberof UsersMiddleware
+      */
+  static validateRegister(req, res, next) {
+    validate(req.body.user, registerSchema).then(() => next()).catch((err) => {
```
This is only really pertinent to registration, so there's no particular reason for it to be a middleware of its own

LordUche replied:
The middleware class is dedicated to validations for separation of concerns and future validation methods will be added to it. This is so that the controller methods are decluttered and perform only one function, *"interact with the database"*

andela-pessien commented on test/users.spec.js:
```diff
@@ -1,7 +1,27 @@
+
+describe('Authentication', () => {
+  describe('POST /api/users', () => {
+    it('should register a user', (done) => {
```
Tests are not very useful if they only test the happy path. Your tests here are expected to cover every situation you've written code to handle.

LordUche replied:
There's a separate story for that on the PT board – [#164139770](https://www.pivotaltracker.com/story/show/164139770)
I intend to write those tests there

andela-pessien replied:
Good work on these by the way :)

andela-pessien commented on server/utils/helpers.js:
```diff
@@ -0,0 +1,57 @@
+ * @param {string | object | Buffer} payload
+ * @returns {string}
+ */
+export const generateToken = payload => jwt.sign(payload, env.SECRET);
```
No expiry?

LordUche replied:
I'll change the function to also accept an `expiresIn` argument.

andela-pessien reviewed (changes requested):
In addition I expect that the issues Hound has raised will also be addressed (not silently resolved).

LordUche committed:
feat(signup): add validation logic to controller

LordUche committed:
feat(signup-validation): signup validation

- write unit tests
- document with swagger
- implement review feedbacks

[Finishes #164139770]

LordUche committed:
feat(signup): fix eslint errors
